### DISCUSSION TITLE
🔒 ERC-20 Decimals Implementation

### DISCUSSION THREAD

**Discussion by jaglinux** [pcaversaccio/snekmate#38] [Ideas]
https://github.com/pcaversaccio/snekmate/blob/main/src/tokens/ERC20.vy#L78 is declared constant.
Is it better to declare it as immutable so that the user can set the value in constructor ? rather than changing the code.
(I guess Vyper does not support virtual / override and hence the decision to have decimals as constant.)

**Comment by jaglinux** [comment]
cc @pcaversaccio

**Comment by pcaversaccio** [comment]
@jaglinux thanks for starting this discussion - when I was writing the code for the `ERC20` contract I thought about it very properly (`immutable` vs. `constant`) and decided to choose `constant` for the following reasons:
- I want to enforce the standard `18` decimals if possible. Decimals can be very confusing and need special care from certain protocols if a token does not adhere to the `18` decimals. Two famous examples that do not conform with the _norm_ of `18` decimals are USDC & USDT. That caused more trouble than good IMHO.
- The most common Solidity library [OpenZeppelin](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol#L87) does default to 18 (a counter-example would be [Solmate](https://github.com/transmissions11/solmate/blob/main/src/tokens/ERC20.sol#L25) which uses an `immutable` variable) even though it's a `virtual` function (a feature that Vyper does not support). So if you want to override it, you must do the following for example:

```solidity
/// @dev Overrides the `decimals` function to use the custom value of 6.
function decimals() public view virtual override returns (uint8) {
    return 6;
}
```

If you do this, you (should) know what you're doing. And the same holds for my `ERC20` version - if you want to implement `decimals` as an `immutable` variable, you can simply do this by adjusting the code. But first, you must understand the code (as well as the implications) and that is what I want to achieve 😀.

**Comment by jaglinux** [comment]
@pcaversaccio Thanks for detailed explanation.
Yeah, I was referring to USDC wherein the decimal is different than 18.
As you mentioned, the OZ implementation requires the user to explicitly override the decimals() and in snekmate implementation the user need to explicitly modify the code. (fair enough)
For the record, [Solmate](https://github.com/transmissions11/solmate/blob/main/src/tokens/ERC20.sol#L25) implements decimals as immutable. It also forces user to specify in the constructor always. If it's gonna be 18 most of the times ,then this can be overhead.

**Reply by pcaversaccio** [reply]
It's pretty simple - if people want to modify the code, they must perfectly understand the code & consequences. And that's what I want to achieve.

**Comment by pcaversaccio** [comment]
UPDATE: I made `decimals` an `immutable` variable via https://github.com/pcaversaccio/snekmate/pull/234, which will be shipped via the `0.1.0` release that targets Vyper version `0.4.0`.
